### PR TITLE
Add smarter getter for block slipperiness

### LINKED ISSUE #1342: Block.slipperiness should have a position sensitive getter
Signature: `(IBlockAccess, int, int, int)float`

Many of the same use cases as #1341. Also only accessed in locations where access to the x/y/z and world exist.

### PR DESCRIPTION
**Pull request by howtonotwin**
Allow a block access to the blockstate, world, position, and entity when getting slipperiness with new `Block::getSlipperiness` method. Change usages of `Block::slipperiness` to `Block::getSlipperiness` in vanilla. Deprecate old field.

[Boat Blaster](https://gist.github.com/howtonotwin/3ff1260ac20d6befbf0784708e30c14f) test mod gist. As per the name, it adds a block (`boat_blaster:boat_blaster`) that blasts boats at high speed (think a hundred blocks/tick) by using the new getter to give the block a slipperiness of `2` for `EntityBoat`s. It retains the default for other entities. (Don't actually be in the boat when this happens. MC complains about the high speed and everything goes choppy and weird. Works fine as long as anything *but* a player is in the boat.)

Resolves #1342.

### CODE REVIEW

**Review by liach — APPROVED** [review]

**Comment by liach** [comment]
You need to fix the merge conflict.

**Comment by bs2609** [comment]
This needs updating again to resolve conflicts @howtonotwin

**Comment by howtonotwin** [comment]
Is there a good, tested way to fix patch conflicts? The way I did it was

1. `git reset --hard HEAD`
2. Checkout `slipperiness` branch
3. `./gradlew setupForge`
4. Copy `projects/Forge/src/main/java` into external location
5. Repeat for `1.11.x` branch and for `git merge-base 1.11.x slipperiness` commit
6. `for file in $(find base -type f -printf '%P\n'); do git merge-file slipperiness/$file base/$file upstream/$file; done`
7. Copy resulting sources back into Forge tree
8. `./gradlew genPatches`

But I'm not sure if this will produce any errors in other cases, or if there's an easier way. I checked the output just in case and it seems fine.

**Comment by mezz** [comment]
No, fixing patch conflicts is not particularly easy. I think you have the right idea. I tend to just fix the code manually and then `genPatches`. This one just happens to touch a pretty popular file, `Block.java`, so it probably conflicts more often than other PRs.

File: patches/minecraft/net/minecraft/block/Block.java.patch [hunk 1]
```diff
@@ -207,6 +215,23 @@
 +    protected ThreadLocal<EntityPlayer> harvesters = new ThreadLocal();
 +    private ThreadLocal<IBlockState> silk_check_state = new ThreadLocal();
 +    protected static java.util.Random RANDOM = new java.util.Random(); // Useful for random things without a seed.
++
++    /**
++     * Gets the slipperiness at the given location at the given state. Normally between 0 and 1.
++     * <p>
++     * Note that entities may reduce slipperiness by a certain factor of their own; for {@link net.minecraft.entity.EntityLivingBase}, this is {@code .91}. {@link net.minecraft.entity.item.EntityItem} uses {@code .98}, and {@link net.minecraft.entity.projectile.EntityFishHook} uses {@code .92}.
```
Comment (liach) [inline patches/minecraft/net/minecraft/block/Block.java.patch]: Can you split this in multiple lines?

**Comment by howtonotwin** [comment]
Should there be a `setDeaultSlipperiness(Float)` so that someone can set the field without generating a deprecation warning? It'll come with a Javadoc message that signifies that actual slipperiness comes from `getSlipperiness` and the set value may not be accurate, or even be wholly ignored.

File: patches/minecraft/net/minecraft/block/Block.java.patch [hunk 2]
```diff
@@ -207,6 +215,25 @@
 +    protected ThreadLocal<EntityPlayer> harvesters = new ThreadLocal();
 +    private ThreadLocal<IBlockState> silk_check_state = new ThreadLocal();
 +    protected static java.util.Random RANDOM = new java.util.Random(); // Useful for random things without a seed.
++
++    /**
++     * Gets the slipperiness at the given location at the given state. Normally between 0 and 1.
++     * <p>
++     * Note that entities may reduce slipperiness by a certain factor of their own; for {@link net.minecraft.entity.EntityLivingBase},
++     * this is {@code .91}. {@link net.minecraft.entity.item.EntityItem} uses {@code .98}, and
++     * {@link net.minecraft.entity.projectile.EntityFishHook} uses {@code .92}.
```
Comment (bs2609) [inline patches/minecraft/net/minecraft/block/Block.java.patch]: Could you split this so each example gets its own line?

Like:
```
 * Note that entities may reduce slipperiness by a certain factor of their own;
 * for {@link net.minecraft.entity.EntityLivingBase}, this is {@code .91},
 * {@link net.minecraft.entity.item.EntityItem} uses {@code .98},
 * and {@link net.minecraft.entity.projectile.EntityFishHook} uses {@code .92}.
```

Comment (howtonotwin) [inline patches/minecraft/net/minecraft/block/Block.java.patch]: How about a table?

**Comment by kashike** [comment]
> Should there be a `setDeaultSlipperiness(Float)` so that someone can set the field without generating a deprecation warning? It'll come with a Javadoc message that signifies that actual slipperiness comes from `getSlipperiness` and the set value may not be accurate, or even be wholly ignored.

Wouldn't hurt, and would prevent complaints about deprecation warnings.

File: patches/minecraft/net/minecraft/block/Block.java.patch [hunk 3]
```diff
@@ -207,6 +215,42 @@
...
++     * <p>
++     * <b>Calling this method may have no effect on the function of this block</b>,
++     * or may not have the expected result. This block is free to caclculate
++     * its slipperiness arbitrarily. This method is guaranteed to work on the
++     * base {@code Block} class.
++     *
++     * @param slipperiness the base slipperiness of this block
++     * @see #getSlipperiness(IBlockState, IBlockAccess, BlockPos, Entity)
++     */
++    public void setDefaultSlipperiness(float slipperiness) {
```
Comment (mezz) [inline patches/minecraft/net/minecraft/block/Block.java.patch]: formatting: braces must be on a new line

**Review by mezz — CHANGES_REQUESTED** [review]
Looks good. 

Please fix formatting I commented on.

Also please include the `BoatBlaster` test mod as `SlipperinessTest` in this PR. It should go with the other forge test mods, and be disabled by default.

**Comment by howtonotwin** [comment]
Should this be rebased to `1.12.x`?

**Comment by mezz** [comment]
Yes, this should target 1.12.x.
If you want to also target 1.11.x, make a new 1.12.x PR and leave this one open.

**Review by mezz — APPROVED** [review]

**Comment by mezz** [comment]
If you want, please update this to match the 1.12 version and I can merge it. Otherwise we should close it.

**Comment by howtonotwin** [comment]
That should be it.

**Pull request merged**